File: algorithms/F/Find Valid Matrix Given Row and Column Sums/Find Valid Matrix Given Row and Column Sums.java
// Greedy Min Row Col Solution
// Time complexity: O(MN)
// Space complexity: O(MN)
class Solution {
    public int[][] restoreMatrix(int[] rowSum, int[] colSum) {
        final int M = rowSum.length, N = colSum.length;
        int[][] matrix = new int[M][N];
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                matrix[i][j] = Math.min(rowSum[i], colSum[j]);
                rowSum[i] -= matrix[i][j];
                colSum[j] -= matrix[i][j];
            }
        }
        return matrix;
    }
}

File: scripts/algorithms/F/Find Valid Matrix Given Row and Column Sums/Find Valid Matrix Given Row and Column Sums.cpp
class Solution {
public:
	vector<vector<int>> restoreMatrix(vector<int>& rowSum, vector<int>& colSum) {
		int nRow = rowSum.size();
		int nCol = colSum.size();

		vector<vector<int>> res(nRow, vector<int> (nCol, 0));

		int i = 0, j = 0;
		while(i < nRow and j < nCol){
			if(rowSum[i] < colSum[j]){
				res[i][j] = rowSum[i];
				colSum[j] -= rowSum[i];
				rowSum[i++] = 0;
			}
			else{
				res[i][j] = colSum[j];
				rowSum[i] -= colSum[j];
				colSum[j++] = 0;
			}
		}
		return res;
	}
};